ext/mbstring/mbstring.c

```diff
@@ -2282,6 +2289,56 @@ PHP_FUNCTION(mb_output_handler)
 }
 /* }}} */
 
+/* {{{ proto array mb_str_split(string str [, int split_length] [, string encoding])
+ Convert a multibyte string to an array. If split_length is specified,
+ break the string down into chunks each split_length characters long. */
+PHP_FUNCTION(mb_str_split)
+{
+	zend_string *str, *encoding = NULL;
+	size_t mblen, offset = 0;
+	mbfl_string string, result, *ret;
+	zend_long split_length = 1;
+
+	ZEND_PARSE_PARAMETERS_START(1, 3)
+		Z_PARAM_STR(str)
+		Z_PARAM_OPTIONAL
+			Z_PARAM_LONG(split_length)
+			Z_PARAM_STR(encoding)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (split_length <= 0) {
+		php_error_docref(NULL, E_WARNING, "The length of each segment must be greater than zero");
+		RETURN_FALSE;
+	}
+
+	/* fill mbfl_string structure */
+	string.no_language = MBSTRG(language);
+	string.encoding = php_mb_get_encoding(encoding);
+	if (!string.encoding) {
+		RETURN_FALSE;
+	}
+
+	string.val = (char unsigned *)ZSTR_VAL(str);
+	string.len = ZSTR_LEN(str);
+
+	/* mb length */
+	mblen = mbfl_strlen(&string);
+
+	/* init array with defined size */
+	array_init_size(return_value, mblen / split_length);
+
+	/* split string */
+	while(offset < mblen){
+		ret = mbfl_substr(&string, &result, offset, (size_t)split_length);
+
+		add_next_index_stringl(return_value, (char const *)ret->val, ret->len);
+
+		efree(ret->val);
+		offset += split_length;
+	}
```

Comment:
While this code will support all encodings, this is also going to be very slow. The issue is that we're operating on codepoint offsets here and finding which byte offset corresponds to a certain codepoint offset is an O(n) operation. Every mbfl_substr call is going to scan the string to find the offset anew. This means that the overall complexity of this function is going to be O(n^2), making it impractically slow for strings beyond a certain length (try to put a string with something like 100000 characters in here -- might also make for a good test).

To make this run in O(n) it will require a more specialized implementation. As already mentioned in https://github.com/php/php-src/pull/3715#issuecomment-449726037, there are usually three different implementations necessary, one for fixed-length encodings (trivial), one for UTF8 (basically what you did initially) and one general (you can keep your current code for the general case for now, as this is less important).

Reply:
done.